File: aoc15/day_18/day_18.ml
let input = Input.char_lists "input" in
let init = Array.make_matrix 100 100 0 in
let () =
  List.iteri
    (fun x row ->
      List.iteri (fun y v -> init.(x).(y) <- (if v = '#' then 1 else 0)) row)
    input
in

let neighs x y =
  List.filter
    (fun (xn, yn) ->
      (xn != x || yn != y) && xn >= 0 && yn >= 0 && xn < 100 && yn < 100)
    (List.flatten
       (List.map
          (fun xs -> List.map (fun ys -> (x + xs, y + ys)) [ -1; 0; 1 ])
          [ -1; 0; 1 ]))
in
let on ls grid =
  List.length (List.filter (fun (x, y) -> grid.(x).(y) = 1) ls)
in

let set_corners grid =
  List.iter
    (fun (x, y) -> grid.(x).(y) <- 1)
    [ (0, 0); (0, 99); (99, 0); (99, 99) ]
in
let step grid corners_on =
  let old = Array.(map copy) grid in
  for x = 0 to 99 do
    for y = 0 to 99 do
      let cnt = on (neighs x y) old in
      grid.(x).(y) <-
        (match (old.(x).(y), cnt) with 1, 2 | 1, 3 | 0, 3 -> 1 | _, _ -> 0)
    done
  done;
  if corners_on then set_corners grid
in
let run corners_on =
  let lights = ref (Array.(map copy) init) in
  for _ = 1 to 100 do
    step !lights corners_on
  done;
  Array.copy !lights
in

let total_on grid =
  Array.fold_left ( + ) 0
    (Array.map (fun row -> Array.fold_left ( + ) 0 row) grid)
in

Printf.printf "First: %d\n" (total_on (run false));
Printf.printf "Second: %d\n" (total_on (run true))

File: aoc15/go/day_18/day_18.go
package main

import (
	"aoc/collections"
	"bufio"
	"fmt"
	"os"
)

func on(row, col int, grid *[100][100]int) int {
	if row >= 0 && row < 100 && col >= 0 && col < 100 {
		return grid[row][col]
	}
	return 0
}

func neighs(row, col int, grid *[100][100]int) int {
	return on(row-1, col-1, grid) +
		on(row-1, col, grid) +
		on(row-1, col+1, grid) +
		on(row, col-1, grid) +
		on(row, col+1, grid) +
		on(row+1, col-1, grid) +
		on(row+1, col, grid) +
		on(row+1, col+1, grid)
}

func run(initGrid [100][100]int, cornersOn bool) int {
	grid := initGrid
	for s := 0; s < 100; s++ {
		newGrid := grid
		for r := 0; r < 100; r++ {
			for c := 0; c < 100; c++ {
				n := neighs(r, c, &grid)
				if grid[r][c] == 1 && n != 2 && n != 3 {
					newGrid[r][c] = 0
				} else if grid[r][c] == 0 && n == 3 {
					newGrid[r][c] = 1
				}
			}
		}
		grid = newGrid
		if cornersOn {
			grid[0][0] = 1
			grid[0][99] = 1
			grid[99][0] = 1
			grid[99][99] = 1
		}
	}

	sum := 0
	for _, row := range grid {
		sum += collections.Sum(row[:])
	}
	return sum
}

func main() {
	input, _ := os.Open("input")
	scanner := bufio.NewScanner(input)

	grid := [100][100]int{}
	r := 0
	for scanner.Scan() {
		line := scanner.Text()
		for c, x := range line {
			if x == '#' {
				grid[r][c] = 1
			}
		}
		r++
	}

	fmt.Println("First:", run(grid, false))
	fmt.Println("Second:", run(grid, true))
}
